### PR TITLE
Add grain_size hyperparameter into Covariance and PCA algorithms

### PR DESCRIPTION
**Pull request by Vika-F**
Following changes were made:

- API for setting and getting the `grain_size` efficiency parameter was added into batch Covariance algorithm
- API for setting and getting `macro_block` and `grain_size` efficiency parameters was added into PCA algorithm with `cov` method

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
Internal Ci failed due to an infrastructure problem.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

not applicable.

### CODE REVIEW

**Comment by Vika-F** [comment]
Re-run CI: http://intel-ci.intel.com/f030b4b9-2fa7-f1c7-ad86-a4bf010d0e2d

File: cpp/daal/src/algorithms/covariance/covariance_impl.i [hunk 1]
```diff
@@ -349,13 +349,18 @@ services::Status updateDenseCrossProductAndSums(bool isNormalized, size_t nFeatu
...
+        DAAL_INT64 numRowsInBlock = getBlockSize<cpu>(nVectors); // number of rows in a data block
+        DAAL_INT64 grainSize      = 1;                           // minimal number of data blocks to be processed by a thread
         if (hyperparameter)
         {
             services::Status status = hyperparameter->find(denseUpdateStepBlockSize, numRowsInBlock);
             DAAL_CHECK_STATUS_VAR(status);
             DAAL_CHECK(0ll < numRowsInBlock, services::ErrorHyperparameterBadValue);
+            status = hyperparameter->find(denseUpdateStepGrainSize, grainSize);
+            DAAL_CHECK_STATUS_VAR(status);
+            DAAL_CHECK(0ll < grainSize, services::ErrorHyperparameterBadValue);
```
Comment (david-cortes-intel) [inline cpp/daal/src/algorithms/covariance/covariance_impl.i:361]: Shouldn't it be <=?

File: cpp/oneapi/dal/algo/covariance/detail/compute_ops.hpp [hunk 2]
```diff
@@ -85,6 +85,8 @@ struct compute_ops {
     void check_parameters_ranges(const param_t& params, const input_t& input) const {
         ONEDAL_ASSERT(params.get_cpu_macro_block() > 0);
         ONEDAL_ASSERT(params.get_cpu_macro_block() <= 0x10000l);
+        ONEDAL_ASSERT(params.get_cpu_grain_size() > 0);
+        ONEDAL_ASSERT(params.get_cpu_grain_size() <= 0x10000l);
```
Comment (david-cortes-intel) [inline cpp/oneapi/dal/algo/covariance/detail/compute_ops.hpp:89]: Why would it need such a limit? If it's higher than the number of CPUs, then it shouldn't matter whether it's `0x10000` or `0x10001`, no? Same for the macro block.

File: cpp/daal/src/algorithms/covariance/covariance_impl.i [hunk 3]
```diff
@@ -349,13 +349,18 @@ services::Status updateDenseCrossProductAndSums(bool isNormalized, size_t nFeatu
         const algorithmFPType nVectorsInv = 1.0 / (double)(nVectors);
 
         /* Split rows by blocks */
-        DAAL_INT64 numRowsInBlock = getBlockSize<cpu>(nVectors);
+        DAAL_INT64 numRowsInBlock = getBlockSize<cpu>(nVectors); // number of rows in a data block
+        DAAL_INT64 grainSize      = 1;                           // minimal number of data blocks to be processed by a thread
```
Comment (david-cortes-intel) [inline cpp/daal/src/algorithms/covariance/covariance_impl.i:353]: Would the final parameters end up being logged somewhere when using verbose mode?

File: cpp/daal/src/algorithms/covariance/covariance_impl.i (hunk 1 above)
Comment (david-cortes-intel) [reply cpp/daal/src/algorithms/covariance/covariance_impl.i:361]: Actually no, realized I read it in reverse. Although it'd be easier to read as `grain_size > 0`.

Comment (Vika-F) [reply cpp/daal/src/algorithms/covariance/covariance_impl.i:361]: I think grainSize cannot be 0.

File: cpp/oneapi/dal/algo/covariance/detail/compute_ops.hpp (hunk 2 above)
Comment (Vika-F) [reply cpp/oneapi/dal/algo/covariance/detail/compute_ops.hpp:89]: This tests that the parameters are more or less meaningful and no garbage is passed to the algorithmic kernel.
But yes, those checks are probably excessive. I can remove those.

File: cpp/oneapi/dal/algo/covariance/backend/cpu/compute_kernel_dense.cpp [hunk 4]
```diff
@@ -14,9 +14,10 @@
 * limitations under the License.
 *******************************************************************************/
 
-#include "daal/src/algorithms/covariance/covariance_kernel.h"
```
Comment (avolkov-intel) [inline cpp/oneapi/dal/algo/covariance/backend/cpu/compute_kernel_dense.cpp:17]: Not necessary

**Review by david-cortes-intel — APPROVED** [review]
LGTM, just a small comment to change the header include from `<>` to `""`.

File: cpp/oneapi/dal/algo/covariance/backend/cpu/compute_kernel_dense.cpp (hunk 4 above)
Comment (Vika-F) [reply cpp/oneapi/dal/algo/covariance/backend/cpu/compute_kernel_dense.cpp:17]: We really need some unification here:
https://github.com/uxlfoundation/oneDAL/blob/main/cpp/oneapi/dal/algo/svm/backend/kernel_function_impl.hpp#L21
https://github.com/uxlfoundation/oneDAL/blob/main/cpp/oneapi/dal/algo/pca/backend/cpu/partial_train_kernel_cov.cpp#L22
https://github.com/uxlfoundation/oneDAL/blob/main/cpp/oneapi/dal/algo/covariance/backend/cpu/finalize_compute_kernel_dense.cpp#L18
...
